src/dtx/dtx_resync.c

```diff
@@ -148,66 +190,65 @@ dtx_status_handle(struct dtx_resync_args *dra)
 
 		rc = dtx_check(dra->po_uuid, cont->sc_uuid,
 			       &dre->dre_dte, layout);
-		if (rc != DTX_ST_COMMITTED && rc != DTX_ST_PREPARED) {
+
+		/* The DTX has been committed (or) ready to be committed on
+		 * some remote replica(s), let's commit the it globally.
+		 */
+		if (rc == DTX_ST_COMMITTED || rc == DTX_ST_PREPARED)
+			goto commit;
+
+		if (rc != -DER_NONEXIST) {
 			/* We are not sure about whether the DTX can be
 			 * committed or not, then we have to skip it.
 			 */
 			D_WARN("Not sure about whether the DTX "DF_UOID
-			       "/"DF_DTI" can be committed or not: %d (1)\n",
+			       "/"DF_DTI" can be committed or not: %d\n",
 			       DP_UOID(dre->dre_oid),
 			       DP_DTI(&dre->dre_xid), rc);
 			dtx_dre_release(drh, dre);
 			continue;
 		}
 
-		/* It is possible that the current DTX becomes committable (in
-		 * CoS cache) or has been committed (on-disk) during above dtx
-		 * check remotely. Let's re-check its state locally.
+		/* Someone (the DTX owner or batched commit ULT) may have
+		 * committed or aborted the DTX during we handling other
+		 * DTXs. So double check the on-disk status before current
+		 * commit.
 		 */
-		rc1 = vos_dtx_check_committable(cont->sc_hdl, &dre->dre_oid,
-					&dre->dre_xid, dre->dre_hash,
-					dre->dre_intent == DAOS_INTENT_PUNCH ?
-					true : false);
-		if (rc1 == DTX_ST_COMMITTED) {
-			/* The DTX is in CoS cache (committable), do nothing. */
+		rc = vos_dtx_check(cont->sc_hdl, &dre->dre_xid);
+
+		/* Skip this DTX since it has been committed or aborted or
+		 * fail to get the status.
+		 */
+		if (rc != DTX_ST_PREPARED) {
+			if (rc < 0 && rc != -DER_NONEXIST)
+				D_WARN("Not sure about whether the DTX "DF_UOID
+				       "/"DF_DTI" can be abort or not: %d\n",
+				       DP_UOID(dre->dre_oid),
+				       DP_DTI(&dre->dre_xid), rc);
 			dtx_dre_release(drh, dre);
 			continue;
 		}
 
-		/* The DTX has been committed on some remote replica(s), let's
-		 * commit the it globally.
-		 */
-		if (rc == DTX_ST_COMMITTED)
-			goto commit;
-
-		switch (rc1) {
-		case DTX_ST_PREPARED:
-			/* Both local and remote replicas are 'prepared', then
-			 * it is committable.
-			 */
-			if (rc == DTX_ST_PREPARED)
-				goto commit;
-
+		rc = vos_dtx_lookup_cos(cont->sc_hdl, &dre->dre_oid,
+					&dre->dre_xid, dre->dre_hash,
+					dre->dre_intent == DAOS_INTENT_PUNCH ?
+					true : false);
```

Comment:
this lookup_cos and dtx_check seems only needed for ongoing dtx on the leader?  But rebuild actually does not care such dtx, a bit confused here. Could you please explain a bit?

Reply:
> this lookup_cos and dtx_check seems only needed for ongoing dtx on the leader?

The leader may be a new leader, it needs to handle all non-committed DTXs leaded by it.

> But rebuild actually does not care such dtx, a bit confused here. Could you please explain a bit?

Any concern?

Reply:
So this is just for handling the new leader case? then we probably can get it from the dtx entry, I remember we store such information? instead of keep checking cos. 

Sorry just trying to understand the code, since it is tightly related with rebuild process.

Reply:
lookup from DTX entry (that is in SCM) is slower than lookup from CoS cache. On the other hand, the DTX entry in SCM maybe 'prepared' status. But if it is SCM, then it is 'committable'.

Reply:
Not sure I follow here, I thought you retrieve the dtx entry from active list (which is from SCM), can we store the leader information there and check.   

Sorry I still am not sure why checking cos and vos_dtx check after collecting the status from all replicas.

If this is for the old leader case, then the check would not provide any benefits, since during the dtx_commit, the dtx might still change to committed, i.e. checking here (same as checking in dtx_commit_resync) would not solve the issue. Or I miss sth.

If this is the new leader case, then the dtx status will not be changed by others except this resync process.  then why do this check.

Sorry, just confused me. Niu, xuezhao any ideas? could you please check this process Thanks.

Reply:
We stored the old leader information in the DTX. But if the old leader is crashed and evicted, we have to elect new leader, and the new leader will do DTX resync.

We load all non-committed DTXs locally, for each of them, the DTX resync ULT may send DTX_CHECK RPC to query with remote replica, then there will be CPU yield. During such internal, some others, such as the batched commit ULT or the DTX owner may commit or abort the DTX. That is the race. So the resync ULT needs double check after the remote check.

Please not that, the leader is not fixed, it is maybe old leader for some DTXs, but new leader for other DTXs. Here, we do not know whether I am the new one or not. The check here is mainly for find out the one to be aborted.

Reply:
If this is the new leader, then why batched commit ULT or other ULT will try to commit this DTX?

Reply:
We do not know whether it is the new leader or old leader... If we want to know that, we must load the DTX firstly, that is the almost the same as vos_dtx_check() does.

Reply:
the dtx is iterated from the active list, which is in SCM, we can store the leader information there to know(see my suggestion above), instead of checking vos_dtx again, which is quite confusing. IMHO.

Reply:
Currently, we have already stored the leader information in the DTX entry, that is added by you. But if there is leader switch, then if we still want to use such information, we have to refresh it, that will introduce addition PMDK overhead, should be avoided.

I am not sure I understand your concern. The logic is as following:

1. The DTX resync logic loads all non-committed DTX into DRAM.

2. For each of above DTX, check whether it is CoS cache or not, it yes, it is the leader, the DTX can be committed.

3. Otherwise, calculate whether it is the new leader or not. If not, skip the DTX. Otherwise, check with remote replica whether the DTX is ready for commit.

4. If the remote replica prepared or committed the DTX, then it is committable.

5. Otherwise, we need to double check local on-disk status to handle the potential race with others. If the on-disk DTX status is committed, then skip the DTX.

6. If the local on-disk DTX status is still 'prepared', then check CoS cache. If it is in the cache, then it is committable, skip it.

7. If the DTX is not in CoS cache, then abort such DTX directly.

8. If the committable DTXs count exceed some count, batched commit them. So before the batched commit, some of the DTX may be committed by race. It is harmless to commit the DTX repeatedly, but not insert into CoS cache.

Reply:
On 4) even all remote replica are prepared, but it does not check the local one, why it can be committable. Especially for changing new leader case.

It seems to me it should check the local dtx status first, then check the remote replica status.

TBH, I think it should split the old leader and new leader case here, 

For old leader, it just iterate the dtx from committable list, then do sync.
For new leader, it can iterate from the active list, then check if this is the new leader by ds_pool_check_leader and the flags in dtx entry.  then do resync in a different path.
Maxing them together make it so confuse here.

is 5) to avoid committing repeatable?  but even you check 5) here, the dtx can still be committed during step 7.  why do 5 then.

6) Not sure why 6. is this for checking invalid DTX? i.e. local dtx is prepared, but remote status are missing?

Reply:
> On 4) even all remote replica are prepared, but it does not check the local one, why it can be committable. Especially for changing new leader case.
> 
> It seems to me it should check the local dtx status first, then check the remote replica status.
> 

This has been done via step 1 when load.

> TBH, I think it should split the old leader and new leader case here,
> 
> For old leader, it just iterate the dtx from committable list, then do sync.
> For new leader, it can iterate from the active list, then check if this is the new leader by ds_pool_check_leader and the flags in dtx entry. then do resync in a different path.
> Maxing them together make it so confuse here.
> 

Even if for the case that no leader switch, the DTX also may be not committable.

> is 5) to avoid committing repeatable? but even you check 5) here, the dtx can still be committed during step 7. why do 5 then.

5) is for avoiding abort the DTX by wrong. Because remote replica may become 'prepared' after replying DTX resync with non-ready.

> 
>  Not sure why 6. is this for checking invalid DTX? i.e. local dtx is prepared, but remote status are missing?

6) is for the case that remote replica reply non-ready, local on-disk status is 'prepared'. Then check whether it is in CoS cache, if there, it is committable, otherwise, it should be aborted.

Reply:
> > On 4) even all remote replica are prepared, but it does not check the local one, why it can be committable. Especially for changing new leader case.
> > It seems to me it should check the local dtx status first, then check the remote replica status.
> 
> This has been done via step 1 when load.

Ah, right.

> 
> > TBH, I think it should split the old leader and new leader case here,
> > For old leader, it just iterate the dtx from committable list, then do sync.
> > For new leader, it can iterate from the active list, then check if this is the new leader by ds_pool_check_leader and the flags in dtx entry. then do resync in a different path.
> > Maxing them together make it so confuse here.
> 
> Even if for the case that no leader switch, the DTX also may be not committable.
> 

For non-leader switch case,  rebuild only care those committable DTX, and hope those will be committed.

> > is 5) to avoid committing repeatable? but even you check 5) here, the dtx can still be committed during step 7. why do 5 then.
> 
> 1. is for avoiding abort the DTX by wrong. Because remote replica may become 'prepared' after replying DTX resync with non-ready.

You mean the race between normal dtx commit and dtx resync?  So why do we care normal dtx commit, I think rebuild does not need those uncommitable DTX?

> 
> > Not sure why 6. is this for checking invalid DTX? i.e. local dtx is prepared, but remote status are missing?
> 
> 1. is for the case that remote replica reply non-ready, local on-disk status is 'prepared'. Then check whether it is in CoS cache, if there, it is committable, otherwise, it should be aborted.

So this is another race case?

Reply:
> > > TBH, I think it should split the old leader and new leader case here,
> > > For old leader, it just iterate the dtx from committable list, then do sync.
> > > For new leader, it can iterate from the active list, then check if this is the new leader by ds_pool_check_leader and the flags in dtx entry. then do resync in a different path.
> > > Maxing them together make it so confuse here.
> > 
> > 
> > Even if for the case that no leader switch, the DTX also may be not committable.
> 
> For non-leader switch case, rebuild only care those committable DTX, and hope those will be committed.
> 

The rebuild will scan the system, if hit non-committed DTX, it will ask the rebuild to retry. That is what we want to avoid. So the old DTX is either been committed, or be aborted.

> > > is 5) to avoid committing repeatable? but even you check 5) here, the dtx can still be committed during step 7. why do 5 then.
> > 
> > 
> > 
> > 1. is for avoiding abort the DTX by wrong. Because remote replica may become 'prepared' after replying DTX resync with non-ready.
> 
> You mean the race between normal dtx commit and dtx resync? So why do we care normal dtx commit, I think rebuild does not need those uncommitable DTX?

In the step 1), the DTX resync ULT find the DTX, it is locally 'prepared', then the ULT asks remote replicas whether it is ready or not, at that time, one replica is not 'prepared', then replies to the leader. But just after the reply, such DTX become 'prepared' on remote replicas, but the DTX resync ULT does not know such change. So the DTX has to double check locally after it gets non-ready reply from remote replicas to avoid aborting committable/committed DTX.

> 
> > > Not sure why 6. is this for checking invalid DTX? i.e. local dtx is prepared, but remote status are missing?
> > 
> > 
> > 
> > 1. is for the case that remote replica reply non-ready, local on-disk status is 'prepared'. Then check whether it is in CoS cache, if there, it is committable, otherwise, it should be aborted.
> 
> So this is another race case?

It is not race. It is normal case that the old leader dispatch modification RPC to non-leaders, but before send to all replicas, it crashed, then the new leader will find that its local has prepared DTX, but some replaces do not have related DTX, then such DTX should be aborted.